### PR TITLE
PLEASE TEST: Upgrade Jellyfin to v10.9

### PR DESCRIPTION
**Pull request by selfhoster1312**
New Jellyfin version ! Still a few TODOs. See next comment.

Many [new features](https://jellyfin.org/posts/jellyfin-release-10.9.0/). Also fixes automatic server discovery (#141).

Some warnings i don't know what to do about:

```
Attention : File /etc/jellyfin/network.xml has been manually modified since the installation or last upgrade. So it has been duplicated in /var/cache/yunohost/appconfbackup//etc/jellyfin/network.xml.backup.20240512.205528
Attention : --- /var/cache/yunohost/appconfbackup//etc/jellyfin/network.xml.backup.20240512.205528      2024-05-12 20:55:00.687566755 +0200
Attention : +++ /etc/jellyfin/network.xml       2024-05-12 20:55:28.335544109 +0200
Attention : @@ -2,35 +2,14 @@
Attention :  <NetworkConfiguration xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
Attention :    <RequireHttps>false</RequireHttps>
Attention :    <BaseUrl>/media</BaseUrl>
Attention : -  <PublicHttpsPort>8920</PublicHttpsPort>
Attention : -  <HttpServerPortNumber>8095</HttpServerPortNumber>
Attention : -  <HttpsPortNumber>8920</HttpsPortNumber>
Attention : +  <InternalHttpPort>8095</InternalHttpPort>
Attention :    <EnableHttps>false</EnableHttps>
Attention : -  <PublicPort>8095</PublicPort>
Attention : -  <UPnPCreateHttpPortMap>false</UPnPCreateHttpPortMap>
Attention : -  <UDPPortRange />
Attention :    <EnableIPV6>true</EnableIPV6>
Attention :    <EnableIPV4>true</EnableIPV4>
Attention : -  <EnableSSDPTracing>false</EnableSSDPTracing>
Attention : -  <SSDPTracingFilter />
Attention : -  <UDPSendCount>2</UDPSendCount>
Attention : -  <UDPSendDelay>100</UDPSendDelay>
Attention : -  <IgnoreVirtualInterfaces>true</IgnoreVirtualInterfaces>
Attention : -  <VirtualInterfaceNames>vEthernet*</VirtualInterfaceNames>
Attention : -  <GatewayMonitorPeriod>60</GatewayMonitorPeriod>
Attention : -  <TrustAllIP6Interfaces>false</TrustAllIP6Interfaces>
Attention : -  <HDHomerunPortRange />
Attention : -  <PublishedServerUriBySubnet>
Attention : -    <string>https://kl.netlib.re/media</string>
Attention : -  </PublishedServerUriBySubnet>
Attention : -  <AutoDiscoveryTracing>false</AutoDiscoveryTracing>
Attention :    <AutoDiscovery>true</AutoDiscovery>
Attention : -  <RemoteIPFilter />
Attention : -  <IsRemoteIPFilterBlacklist>false</IsRemoteIPFilterBlacklist>
Attention :    <EnableUPnP>true</EnableUPnP>
Attention :    <EnableRemoteAccess>true</EnableRemoteAccess>
Attention : -  <LocalNetworkSubnets />
Attention : -  <LocalNetworkAddresses />
Attention : -  <KnownProxies />
Attention : -  <EnablePublishedServerUriByRequest>false</EnablePublishedServerUriByRequest>
Attention : -</NetworkConfiguration>
Attention : \ Pas de fin de ligne à la fin du fichier
Attention : +  <LocalNetworkAddresses>
Attention : +    <string>127.0.0.1</string>
Attention : +  </LocalNetworkAddresses>
Attention : +</NetworkConfiguration>
```

Some warnings i added to move old plugins directory instead of accumulating dust:

```
Attention : Jellyfin plugin 'LDAP Authentication' has different ABI version 10.8.0.0 than expected 10.9.0.0. Moving to /var/lib/jellyfin/plugins-backup.
Attention : Jellyfin plugin 'LDAP Authentication_12.0.0.0' has different ABI version 10.7.0.0 than expected 10.9.0.0. Moving to /var/lib/jellyfin/plugins-backup.
Attention : Jellyfin plugin 'Open Subtitles_10.0.0.0' has different ABI version 10.7.0.0 than expected 10.9.0.0. Moving to /var/lib/jellyfin/plugins-backup.
Attention : Jellyfin plugin 'Open Subtitles_11.0.0.0' has different ABI version 10.7.0.0 than expected 10.9.0.0. Moving to /var/lib/jellyfin/plugins-backup.
Attention : Jellyfin plugin 'Open Subtitles_19.0.0.0' has different ABI version 10.8.10.0 than expected 10.9.0.0. Moving to /var/lib/jellyfin/plugins-backup.
```

### CODE REVIEW

**Comment by selfhoster1312** [comment]
TODO:

- respect the old `manifest.resources.sources` keys
- delete the changed config warnings
- investigate why it takes two restart/daemon-reload/restart to take the new environment published URL into effect
- update documentation

**Comment by selfhoster1312** [comment]
If someone has an idea why the extra systemd config is only taken into consideration after a second daemon-reload/restart [here](https://github.com/YunoHost-Apps/jellyfin_ynh/pull/154/files#diff-57aeb84da86cb7420dfedd8e49bc644fb799d5413d01927a0417bde753e8922fR207) ?

**PLEASE TEST AND REPORT ISSUES**

**Comment by ericgaspar** [comment]
!testme

File: manifest.toml [hunk 1]
```diff
@@ -56,81 +56,97 @@ ram.runtime = "100M"
...
-
-
-    [resources.sources.plugin_ldap]
-    url = "https://repo.jellyfin.org/releases/plugin/ldap-authentication/ldap-authentication_17.0.0.0.zip"
-    sha256 = "042bdd16950c7569c154311cc64af49c7d6096e9b0a7184287707d5ee317ecf5"
-    in_subdir = false
-
-
+[resources.sources]
+format = "whatever"
```
Comment (Salamandar) [inline manifest.toml:60]: Having values here is invalid.

**Review by Salamandar — CHANGES_REQUESTED** [review]

File: scripts/upgrade [hunk 2]
```diff
@@ -202,11 +202,7 @@ systemctl daemon-reload
 #=================================================
 ynh_script_progression --message="Starting a systemd service..." --weight=1
 
-ynh_systemd_action --service_name=$app --action="start" --log_path="systemd" --timeout=15
-
-# Why do i have to do this again to pickup the new systemd config???
-systemctl daemon-reload
-systemctl restart jellyfin
+ynh_systemd_action --service_name=$app --action="restart" --log_path="systemd" --timeout=15
```
Comment (selfhoster1312) [inline scripts/upgrade:205]: Not sure that's enough. It wasn't picking up the new systemd override with just one call. I'll test tomorrow and report back.

File: scripts/upgrade [hunk 3]
```diff
@@ -182,7 +182,7 @@ ynh_multimedia_addaccess $app
 ynh_script_progression --message="Upgrading logrotate configuration..." --weight=1
 
 # Use logrotate to manage app-specific logfile(s)
-ynh_use_logrotate
+ynh_use_logrotate --non-append
```
Comment (selfhoster1312) [inline scripts/upgrade:185]: Apparently this option is deprecated. It was failing at some point. see apps chatroom log for more detail

File: scripts/_common.sh [hunk 4]
```diff
@@ -37,16 +37,28 @@ install_jellyfin_packages() {
...
 	ynh_package_install \
-		"${tempdir}/jellyfin-web.deb" \
-		"${tempdir}/jellyfin-server.deb" \
-		"${tempdir}/jellyfin-ffmpeg.deb"
+		"${tempdir}/jellyfin-ffmpeg6.deb"
+
+	# The doc says it should be called only once,
+	# but the code says multiple calls are supported.
+	# Also, they're already installed so that should be quasi instantaneous.
+	ynh_install_app_dependencies \
```
Comment (selfhoster1312) [inline scripts/_common.sh:58]: Why 2 install steps? Can we not just choose one?  It was working with just one...

File: scripts/upgrade (hunk 2 above)
Comment (selfhoster1312) [reply scripts/upgrade:205]: Its enough from initial testing

**Comment by selfhoster1312** [comment]
!testme

**Comment by selfhoster1312** [comment]
!testme

**Comment by Salamandar** [comment]
>     * [ ]  investigate why it takes two restart/daemon-reload/restart to take the new environment published URL into effect

It looks like it's more about startup delay. At least on my end with a single `restart` it works but not right away.

File: conf/systemd.service [hunk 5]
```diff
@@ -0,0 +1,2 @@
+[Service]
+Environment="JELLYFIN_PublishedServerUrl=https://__DOMAIN____PATH__"
```
Comment (Salamandar) [inline conf/systemd.service:2]: Same here, what's the doc for that ? It looks like it's related to docker based installs.

File: conf/network.xml
[the remark this answers is not in the document]
Comment (selfhoster1312) [reply conf/network.xml:1]: Upstream says [they don't document setting files](https://github.com/jellyfin/jellyfin/issues/11564#issuecomment-2106249866). I took previous Yunohost configuration, noticed some settings were not applied (Jellyfin was listening on LAN not just localhost), read source code to find out what settings did not exist anymore. Removed stuff until what "made sense" was left.

File: scripts/upgrade [hunk 6]
```diff
@@ -163,7 +182,7 @@ ynh_multimedia_addaccess $app
 ynh_script_progression --message="Upgrading logrotate configuration..." --weight=1
 
 # Use logrotate to manage app-specific logfile(s)
-ynh_use_logrotate --non-append
+ynh_use_logrotate
```
[the remark this answers is not in the document]
Comment (selfhoster1312) [reply scripts/upgrade:185]: It's deprecated

File: scripts/_common.sh (hunk 4 above)
Comment (selfhoster1312) [reply scripts/_common.sh:59]: Apparently they don't do the same stuff and we should leave it like that

File: conf/systemd.service (hunk 5 above)
Comment (selfhoster1312) [reply conf/systemd.service:2]: It's the config to fix #141 

I wanted to add it to network.xml but upstream only wants ENV/CLI setting for this, no configuration file option, see https://github.com/jellyfin/jellyfin/issues/11564

File: conf/network.xml
[the remark this answers is not in the document]
Comment (Salamandar) [reply conf/network.xml:1]: Thanks for the link. I added a reply for clarification on their side.

File: conf/systemd.service (hunk 5 above)
Comment (Salamandar) [reply conf/systemd.service:2]: Alright. I finally found the doc : https://jellyfin.org/docs/general/administration/configuration#main-configuration

**Comment by selfhoster1312** [comment]
> > ```
> > * [ ]  investigate why it takes two restart/daemon-reload/restart to take the new environment published URL into effect
> > ```
> 
> It looks like it's more about startup delay. At least on my end with a single `restart` it works but not right away.

No it's not. Before i was testing with daemon-reload/start it was just NOT working. After i restarted second time it was working. Systematically, not randomly. But apparently 1 restart is enough so that's good.

**Comment by selfhoster1312** [comment]
So we have a few tests failing and a few warnings.

A warning about logging.json not being in same state. Why?

```
26072 WARNING Configuration file '/etc/jellyfin/logging.json'
26073 WARNING  ==> File on system created by you or by a script.
26073 WARNING  ==> File also in package provided by package maintainer.
26073 WARNING  ==> Keeping old config file as default.
```

[Upgrade test on bookworm fails](https://ci-apps-bookworm-dev.yunohost.org/ci/job/959) because the app was not supported on bookworm before... so that's good?

```
1504 WARNING Unknown release: bookworm
1606 ERROR   Unable to install jellyfin: An error occurred inside the app installation script
```

[Upgrade test on bullseye fails](https://ci-apps-dev.yunohost.org/ci/job/15874) because old packages are no longer available... so not good?

```
59748 WARNING https://repo.jellyfin.org/releases/server/debian/versions/jellyfin-ffmpeg/5.0.1-5/jellyfin-ffmpeg5_5.0.1-5-bullseye_amd64.deb:
59749 WARNING 2024-05-14 12:03:59 ERROR 404: Not found.
61252 ERROR Unable to install jellyfin: An error occurred inside the app installation script
```

**Comment by selfhoster1312** [comment]
!testme

**Comment by selfhoster1312** [comment]
!testme

**Comment by selfhoster1312** [comment]
Soooo in addition to the failing upgrade test because old version can't be installed... we have a few warnings:

> Using helper ynh_permission_update is deprecated when using packaging v2 ... It is replaced by: the permission resource

But there is no way to set the admin permission to the $admin setting?

> Using helper ynh_remove_app_dependencies is deprecated when using packaging v2 ... It is replaced by: the apt resource 

But apparently it's ok because we install DEB manually so it's a false warning?

> Using helper ynh_system_user_create is deprecated when using packaging v2 ... It is replaced by: the system_user resource 

But there's no way we can add the $app user to group render via manifest.toml so it's ok?

There's still one warning to deal with:

```
19778 WARNING Configuration file '/etc/jellyfin/logging.json'
19778 WARNING  ==> File on system created by you or by a script.
19778 WARNING  ==> File also in package provided by package maintainer.
19779 WARNING  ==> Keeping old config file as default.
```

**Comment by selfhoster1312** [comment]
!testme

**Comment by nbu2209** [comment]
Hey guys any news on that update ?

**Comment by selfhoster1312** [comment]
> Hey guys any news on that update ?

Still a few warnings i need help to fix :)

But i'm running 10.9.5 on my server and working great

**Comment by stilobique** [comment]
Hello @selfhoster1312 
I made an upgrade with your branch, no issue with the install and the service are functional ; however, I have a message with the log.

```txt
juin 09 10:02:55 systemd[1]: jellyfin.service: Consumed 1d 18h 29min 35.221s CPU time.
juin 09 10:03:42 systemd[1]: Started Jellyfin Media Server.
juin 09 10:03:51 jellyfin[725476]: [10:03:51] [WRN] [1] Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware: The WebRootPath was not found: /var/lib/jellyfin/wwwroot. Static files may be unavailable.
juin 09 10:03:56 jellyfin[725476]: [10:03:56] [ERR] [15] MediaBrowser.Providers.Manager.ProviderManager: Error refreshing item
juin 09 10:03:56 jellyfin[725476]: System.ArgumentException: Guid can't be empty (Parameter 'id')
juin 09 10:03:56 jellyfin[725476]:    at Emby.Server.Implementations.Library.LibraryManager.GetItemById(Guid id)
juin 09 10:03:56 jellyfin[725476]:    at MediaBrowser.Providers.Manager.ProviderManager.StartProcessingRefreshQueue()
```

By the way, everything works ; thanks for your job ;) .

**Comment by tituspijean** [comment]
I have just installed the upgrade with no issue (https://paste.yunohost.org/raw/neyuhuxuye):
- I've got the same "The WebRootPath was not found" warning in the jellyfin service log
- No error regarding empty Guid
- I had to force-refresh the browser cache for the web ui
- No issue on Jellyfin for my LG Android TV (though I did not try to disconnect and reconnect)
(sorry stilobique 😅)

**Comment by tituspijean** [comment]
@retiolus I guess it is coming from the deb package deployment, but can you share the upgrade log? Maybe check upstream repo for that file?

**Comment by selfhoster1312** [comment]
Updated to 10.9.6

**Comment by rmkni** [comment]
Hi there, I tried to install from this PR on a fresh yunohost and got this issue :
```
2024-06-20 14:01:59,970: INFO - [###############+....] > Integrating service in YunoHost...
2024-06-20 14:01:59,975: WARNING - ./install: ligne 111: ports : variable sans liaison
2024-06-20 14:01:59,976: DEBUG - ++ ynh_exit_properly
2024-06-20 14:01:59,976: DEBUG - ++ [[ install =~ ^install$|^upgrade$|^restore$ ]]
2024-06-20 14:02:00,492: DEBUG - ++ exit 1
2024-06-20 14:02:01,497: ERROR - Impossible d'installer jellyfin : Une erreur est survenue dans le script d'installation de l'application
```
Maybe there is an error in [install line 111](https://github.com/selfhoster1312/jellyfin_ynh/blob/78ece8d6183c2e140401c270abf68c9d462a0ddc/scripts/install#L111)
I'm not sure about this ${ports} variable

**Comment by selfhoster1312** [comment]
!testme

**Comment by selfhoster1312** [comment]
!testme

**Comment by kay0u** [comment]
!testme

**Comment by kay0u** [comment]
!testme

**Comment by kay0u** [comment]
!testme

**Comment by kay0u** [comment]
!testme

File: scripts/install [hunk 7]
```diff
@@ -106,10 +105,19 @@ ynh_use_logrotate
 #=================================================
 ynh_script_progression --message="Integrating service in YunoHost..." --weight=1
 
-yunohost service add "$app" --description="Jellyfin media center"
+ports_args=()
+if [[ "${opened_ports:-__NOTHING__}" != "__NOTHING__" ]]; then
```
Comment (Salamandar) [inline scripts/install:109]: FYI you can also use `-z` to check the emptiness of the string. Your code is completely fine with me though, don't change it.

**Comment by AltruisticRambler** [comment]
hey all -- not to gum up the comments on this thread but I wanted to toss some thanks in here! I was a bit bummed to find that Jellyfin didn't install earlier today, but was relieved when I found this thread! 

Ya'll rule for getting this app working - stoked to see this get pushed to the main branch! 😃 Cheers!

**Comment by kay0u** [comment]
Merging in testing branch soon

**Pull request merged**